[Driver] Add support for GCC installation detection in Baremetal toolchain

quic-garvgupt opened the pull request:
This patch introduces enhancements to the Baremetal toolchain to support GCC toolchain detection.
- If the --gcc-install-dir or --gcc-toolchain options are provided and point to valid paths, the sysroot is derived from those locations.
- If not, the logic falls back to the existing sysroot inference mechanism already present in the Baremetal toolchain.
- Support for adding include paths for the libstdc++ library has also been added.

Additionally, the restriction to always use the integrated assembler has been removed. With a valid GCC installation, the GNU assembler can now be used as well.

This patch currently updates and adds tests for the ARM target only. RISC-V-specific tests will be introduced in a later patch, once the RISCVToolChain is fully merged into the Baremetal toolchain. At this stage, there is no way to test the RISC-V target within this PR.

RFC: https://discourse.llvm.org/t/merging-riscvtoolchain-and-baremetal-toolchains/75524

Commit:
[Driver] Teach Barmetal toolchain about GCC installation

This patch introduces the baretmetal toolchain object about GCC Installation.
Currently, if `--gcc-installation` ot `--gcc-install-dir` options are passed on
commandline, then sysroot will be formed from there if paths will be valid.
Otherwise, it will be fallback to as it already existed in the Baremetal
toolchaibn object. Moreover, support for adding include paths for libstd
C++ library is added as well.

Additionally, the restriction to always use integrated assembler is removed
because with valid gcc installation, gnu assembler can be invoked as well.

This patch currently adds and modifies arm related test only. The riscv specific
test will be added in the last PR when driver code related to calling of
RISCVToolchain object will be removed. Currently in this PR, there is no way to
test riscv target.

RFC:
https://discourse.llvm.org/t/merging-riscvtoolchain-and-baremetal-toolchains/75524

Change-Id: Ibaeb569cf7e2cee03c022aa9ecd1abe29d5c30d4

quic-garvgupt commented:
* **#121829** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/121829?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/121829?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

quic-garvgupt commented on clang/lib/Driver/ToolChains/BareMetal.cpp:
```diff
@@ -110,20 +111,93 @@ static std::string computeBaseSysRoot(const Driver &D, bool IncludeTriple) {
+                            getDriver().getTargetTriple());
+  }
+
+  if (!llvm::sys::fs::exists(SysRootDir))
```
I understand that. However, to preserve the behavior of both the BareMetal and RISCVToolchain objects, I need to maintain this condition. If this returns empty, only then does the control transfer to compute the sysroot, as it was done previously in the BareMetal toolchain. (See line 152 under `BareMetal::computeSysRoot()` function)

MaskRay replied:
This divergence is ok if eventually the condition will be removed. We do want to drop this difference in the future...

quic-garvgupt replied:
I agree that the condition can be removed if it is actually not needed. Though that has to be done as a part of new patch because for now, this patch aims to preserve the behavior as it was before.

MaskRay replied:
Thanks. If there is commitment to remove the quirks, this looks good to me!

petrhosek replied:
I believe the underlying issue is that BareMetal and RISCVToolchain original authors made different choices when it comes to the default sysroot location:

* BareMetal uses the `bin/../lib/clang-runtimes/<TargetTriple>` directory.
* RISCVToolchain uses the `bin/../<TargetTriple>` directory.

Unless we deprecate one (or both) of these and force everyone to switch to a common one, I don't think we can avoid the directory check.

I'd be in favor of unifying on a single location sooner rather than later even if it's going to require a migration for some users. This would be probably best discussed in an RFC though.

quic-garvgupt commented:
Hi @petrhosek, following up on our last RISC-V embedded sync-up, can you please review this patch? Thanks

quic-garvgupt commented:
It's been a few weeks since this patch was last reviewed. If everything looks good, could someone please provide an LGTM? I'd like to merge this patch soon.

petrhosek commented on clang/lib/Driver/ToolChains/BareMetal.cpp:
```diff
@@ -110,20 +111,93 @@ static std::string computeBaseSysRoot(const Driver &D, bool IncludeTriple) {
+    return SysRoot;
+
+  std::string SysRoot = getDriver().SysRoot;
+  if (!SysRoot.empty() && llvm::sys::fs::exists(SysRoot))
```
We don't check if the directory exists in other drivers so I don't think we should do it here for consistency.
```suggestion
  if (!SysRoot.empty())
```

quic-garvgupt replied:
Done

petrhosek commented on clang/lib/Driver/ToolChains/BareMetal.cpp:
```diff
@@ -110,20 +111,93 @@ static std::string computeBaseSysRoot(const Driver &D, bool IncludeTriple) {
+    StringRef TripleStr = GCCInstallation.getTriple().str();
+    llvm::sys::path::append(SysRootDir, LibDir, "..", TripleStr);
+  } else {
+    // Use the triple as provided to the driver. Unlike the parsed triple
+    // this has not been normalized to always contain every field.
+    llvm::sys::path::append(SysRootDir, getDriver().Dir, "..",
+                            getDriver().getTargetTriple());
```
I understand that this is copied as-is from `RISCVToolchain`, but it's quite confusing for this to be inside the method called `computeGCCSysroot` when it is not in fact a GCC sysroot, it's the sysroot installed alongside Clang.

quic-garvgupt replied:
Inlined the computeGCCSysroot inside the computeSysRoot function.

petrhosek commented on clang/lib/Driver/ToolChains/BareMetal.cpp:
```diff
@@ -97,7 +97,8 @@ static bool findRISCVMultilibs(const Driver &D,
 }
 
-static std::string computeBaseSysRoot(const Driver &D, bool IncludeTriple) {
+static std::string computeInstalledToolchainSysRoot(const Driver &D,
```
Can we avoid renaming this method? That doesn't seem to be necessary and it only creates unnecessary churn.

quic-garvgupt replied:
Done

petrhosek commented on clang/lib/Driver/ToolChains/BareMetal.cpp:
```diff
@@ -110,20 +111,93 @@ static std::string computeBaseSysRoot(const Driver &D, bool IncludeTriple) {
+// bin/../<TargetTriple> directory if it is not explicitly specified on the
+// command line through `--sysroot` option. libc here will be newlib.
+std::string BareMetal::computeGCCSysRoot() const {
+  if (!getDriver().SysRoot.empty())
+    return getDriver().SysRoot;
```
This check shouldn't be necessary.

quic-garvgupt replied:
Removed the check. Thanks for pointing this out!

quic-garvgupt commented:
Hi @petrhosek, I've addressed all your comments. Please review the changes and approve the PR if everything looks good.

petrhosek commented on clang/lib/Driver/ToolChains/BareMetal.cpp:
```diff
@@ -110,20 +110,76 @@ static std::string computeBaseSysRoot(const Driver &D, bool IncludeTriple) {
-      getLibraryPaths().push_back(std::string(Dir));
+    : Generic_ELF(D, Triple, Args) {
+  GCCInstallation.init(Triple, Args);
+  SysRoot = computeSysRoot();
```
I'd make this a local variable.
```suggestion
  std::string SysRoot = computeSysRoot();
```

petrhosek commented on clang/lib/Driver/ToolChains/BareMetal.h:
```diff
@@ -79,7 +87,6 @@ class LLVM_LIBRARY_VISIBILITY BareMetal : public ToolChain {
   OrderedMultilibs getOrderedMultilibs() const;
 
   std::string SysRoot;
```
I'd remove this class member to avoid confusion whether you're referring to a local variable or a class member in methods.
```suggestion
```

quic-garvgupt replied:
Removing this private variable was making the code messy because there have been updates to this variable in some member functions and the updated value is then further used in other functions.  Removing this variable would need to pass this as an argument to all those functions which would make the code less readable.

To avoid the confusion, all local variables for sysroot are named as `SysRootDir` and any references to this variable as `SysRoot` only

petrhosek commented on clang/lib/Driver/ToolChains/BareMetal.cpp:
```diff
@@ -110,20 +110,76 @@ static std::string computeBaseSysRoot(const Driver &D, bool IncludeTriple) {
+
+  if (llvm::sys::fs::exists(SysRootDir))
+    return std::string(SysRootDir);
+  SysRoot = computeBaseSysRoot(getDriver(), /*IncludeTriple*/ true);
+
+  return SysRoot;
```
This can be combined into a single line.
```suggestion
  return computeBaseSysRoot(getDriver(), /*IncludeTriple*/ true);
```

quic-garvgupt replied:
Thanks for pointing this out!. I have fixed it in the next patchset

quic-garvgupt commented:
Hi @petrhosek, I've addressed all your comments. Please review the changes and approve the PR if everything looks good.

quic-garvgupt commented:
Gentle Ping again!

smithp35 commented on clang/lib/Driver/ToolChains/BareMetal.cpp:
```diff
@@ -110,20 +110,81 @@ static std::string computeBaseSysRoot(const Driver &D, bool IncludeTriple) {
   return std::string(SysRootDir);
 }
 
+static bool hasGCCToolChainAlongSideClang(const Driver &D) {
```
IIUC this is when we have something like clang copied into a GNU installation bin directory or another directory at the same level. So that `bin\clang\..\<target-triple>\lib\crt0.o`

One minor concern is that almost any toolchain will have a `crt0.o` although in the current bare-metal driver they are most likely in `clang-runtimes` so won't clash with the location.

I'm wondering if there's a common file that is really GCC specific so we can avoid inferring a GNU toolchain by mistake.

We could decide that this is unlikely, or make it a requirment that non GNU based bare-metal toolchains just avoid this. If that's the case could we document this? I'll leave a separate comment surrounding documentation.

petrhosek replied:
If understand correctly, this should cover the case when we're using newlib (and potentially libgcc and libstdc++) together with Clang (but not using `clang-runtimes`). @quic-garvgupt can you clarify this? If that's the case, then I think this function could use a better name. In either case, I agree with @smithp35 that this logic is fragile since `crt0.o` is commonly used by most C libraries.

smithp35 replied:
> I am not aware if there is any common file that is really GCC specific. Pls do let me know if there is such a file or any other feedback/comments on how this can be handled better in general.

I don't think there's a single common file that is unique to GCC.

Stepping back a bit, there's a bit of the logic/use-case in `computeSysRoot` that I don't understand. When we find a GCC ToolChain alongside clang, I don't see any logic to validate it. There is code later on in `BareMetal::BareMetal` to use the GNU multilib detection only if the GCC installation is valid.

Is the intention here to find a valid GCC installation equivalent to the  `-gcc-install-dir` or `--gcc-toolchain`, or something else? For example this could be an equivalent to `lib/clang-runtimes` although in that case it wouldn't necessarily be a GCC installation.

If the intention is to find a valid installation then can we borrow some of the logic from the `GCCInstallation`?

smithp35 commented on clang/test/Driver/arm-gnutools.c:
```diff
@@ -0,0 +1,12 @@
+// check that gnu assembler is invoked with arm baremetal as well
+
+// RUN: %clang --target=armv6m-none-eabi  --gcc-toolchain=%S/Inputs/basic_riscv32_tree -fno-integrated-as %s -### -c \
```
I'm a bit confused about this test. It is called `arm-gnutools.c`, but has both arm and aarch64 triples, and uses a `basic_riscv32 tree`.

Is it just that we're using `basic_riscv32_tree` to tell the driver that it is in GNU "mode" so look for `<triple>-as` 

Could be worth a comment if that's the case. There's a possibility that the driver could actually look for the exectuable in the GCC toolchain.

quic-garvgupt replied:
Thanks for catching this!. I fixed this in some previous version directly on this branch through which I push however didn't do it on my development branch where I do all my experiments. Therefore while making further changes on this PR I still ended up pushing the incorrect version. In the current version, I am using the correct gcc-toolchain path and armv6m and aarch64 tests are in different files.

smithp35 commented on clang/lib/Driver/ToolChains/BareMetal.cpp:
```diff
@@ -110,20 +110,81 @@ static std::string computeBaseSysRoot(const Driver &D, bool IncludeTriple) {
+  SysRoot = computeSysRoot();
+  if (GCCInstallation.isValid()) {
+    Multilibs = GCCInstallation.getMultilibs();
+    SelectedMultilibs.assign({GCCInstallation.getMultilib()});
```
Are we delegating GCC multilibs entirely to the GNU toolchain driver? As I understand it these are hard-coded at the moment, so I expect the GNU arm-none-eabi toolchains won't work unless their multilibs are added.

Have I got that right? Or will this only work with toolchains, possibly RISCV, that already have their multilib structure encoded in GCCInstallation.

quic-garvgupt replied:
Yes, GNU arm-none-eabi toolchains won't work since mutlilibs for this target is not added. Currently, only form some targets hard-coded multilibs are added including RISCV. One can look at `ScanGCCForMultilibs` function in `Gnu.cpp` to see for what all architectures GCC multilibs are supported IINW.

smithp35 replied:
Is there a chance of an unsupported warning message if the Driver doesn't have the logic implemented yet?  

The bare metal driver supports ARM, AArch64, RISCV and PPC, if multilibs only works on RISCV (PPC looks unlikely), then we could save people some time. I can remember  wasting a lot of time thinking I had got my configuration options wrong due to the generic "not found" error messages, when the driver was never going to work whatever paths I put in..

quic-garvgupt replied:
Yes, I have added a warning in the latest patchset.

smithp35 reviewed (commented):
I'm looking at this patch in isolation, so its possible that the comments here are addressed elasewhere.

It looks like useful functionality to add to the bare metal driver.

It also looks like there is some scope for some documentation, for users and for toolchain developers. Users will need to know about what options to set to point at their GCC installation (or place adjacent). Non-GNU toolchain developers will need to avoid `<triple>/lib/crt0.o`. If multilibs doesn't work for all toolchains yet, would be good to add that as a limitation.

That could be done with a follow up patch though.

quic-garvgupt commented:
> I'm looking at this patch in isolation, so its possible that the comments here are addressed elasewhere.
> 
> It looks like useful functionality to add to the bare metal driver.
> 
> It also looks like there is some scope for some documentation, for users and for toolchain developers. Users will need to know about what options to set to point at their GCC installation (or place adjacent). Non-GNU toolchain developers will need to avoid `<triple>/lib/crt0.o`. If multilibs doesn't work for all toolchains yet, would be good to add that as a limitation.
> 
> That could be done with a follow up patch though.

Thanks for providing feedback @smithp35. I can add documentation related to the points you mentioned. Pls let me know if adding them in the BareMetal.cpp file would be ok or should I consider adding it in some other file.

smithp35 commented on clang/lib/Driver/ToolChains/BareMetal.cpp:
```diff
@@ -110,20 +110,81 @@ static std::string computeBaseSysRoot(const Driver &D, bool IncludeTriple) {
+                         .str());
+    PPaths.push_back((GCCInstallation.getParentLibPath() + "/../bin").str());
+    getFilePaths().push_back(SysRoot + "/lib");
+  } else {
```
IIUC we hit this point when in `BareMetal::computeSysRoot()` when the GCCInstallation isn't valid, but there is a GCC Toolchain adjacent 
```
if (GCCInstallation.isValid()) {
  ...
}
else if (hasGCCToolChainAlongSideClang(D)) {
    // Use the triple as provided to the driver. Unlike the parsed triple
    // this has not been normalized to always contain every field.
    llvm::sys::path::append(SysRootDir, D.Dir, "..", D.getTargetTriple());
}
```

The logic in this block is the same as if clang-runtimes with multilib.yaml were used. Is this intentional?

quic-garvgupt replied:
> Stepping back a bit, there's a bit of the logic/use-case in computeSysRoot that I don't understand. When we find a GCC ToolChain alongside clang, I don't see any logic to validate it.
 
There is no code to validate it in the constructor. However, it's use was needed in the linker job (see `Linker::ConstructJob`). This was done to preserve the behavior of RISCVToolChain Object. I am not sure why it was decided there that the mere presence of `lib/crt0.o` file alongside clang in target triple prefixed directory will indicate a GCC toolchain, however in RISCVToolchain object, toolchain found along clang  isn't considered a valid GCCInstallation and the same has been implemented here to preserve that behavior. 
One needs to look at `hasGCCToolchain`, `computeSysRoot` and constructor in RISCVToolChain.cpp for the same
 
>Is the intention here to find a valid GCC installation equivalent to the -gcc-install-dir or --gcc-toolchain, or something else?
 
Yes it is trying to find a  toolchain that is equivalent to `--gcc-toolchain`. The same is also mentioned in the comments for the function `hasGCCToolChain` in RISCVToolChain.cpp. You can also look at tests "clang/test/Driver/riscv64-toolchain-extra.c` and "clang/test/Driver/riscv32-toolchain-extra.c" to see how it works.
 
> The logic in this block is the same as if clang-runtimes with multilib.yaml were used. Is this intentional?
 
I am not sure if I understood your question completely. But yes, the logic is a bit similar but the directory for sysroot will be different. In the case when GCCInstallation isn't valid but there is a GCC toolchain adjacent, then the sysroot would be `bin/../<target-triple>/lib/crt0.o` whereas with clang-runtimes with multilib.yaml, it will be `bin/../lib/clang-runtimes/` (include triple is set to false, when multilib.yaml is found). 

Apologies for the late reply. I was out for sometime in the last month and then I my commit access got revoked. Thanks for reviewing this patch and pls let me know if you have further questions/comments.

smithp35 commented:
> > I'm looking at this patch in isolation, so its possible that the comments here are addressed elasewhere.
> > It looks like useful functionality to add to the bare metal driver.
> > It also looks like there is some scope for some documentation, for users and for toolchain developers. Users will need to know about what options to set to point at their GCC installation (or place adjacent). Non-GNU toolchain developers will need to avoid `<triple>/lib/crt0.o`. If multilibs doesn't work for all toolchains yet, would be good to add that as a limitation.
> > That could be done with a follow up patch though.
> 
> Thanks for providing feedback @smithp35. I can add documentation related to the points you mentioned. Pls let me know if adding them in the BareMetal.cpp file would be ok or should I consider adding it in some other file.

I think developer documentation in BareMetal.cpp is fine. User documentation is probably best in the clang/docs, for example updating https://github.com/llvm/llvm-project/blob/main/clang/docs/Toolchain.rst

smithp35 reviewed (commented):
Apologies I've got a few more questions as I'm not sure I've understood.

quic-garvgupt commented:
Hi @smithp35, can you pls review the latest changes and approve this PR if everything looks fine?

smithp35 reviewed (commented):
I've done another review pass. Mostly some small stylistic comments.

I'm finding it difficult to follow the code with the various different code-paths with similar names. I've made some suggestions on renaming. We may also be able to add some more comments to make it clearer.

quic-garvgupt commented on clang/lib/Driver/ToolChains/BareMetal.cpp:
```diff
@@ -110,56 +144,99 @@ static std::string computeBaseSysRoot(const Driver &D, bool IncludeTriple) {
-      Triple.getArch() != llvm::Triple::aarch64_be)
-    return false;
+// Users can specify their GCC toolchain using `-gcc-install-dir` or
+// `--gcc-toolchain`. If no sysroot is explicitly provided, the driver will
```
Done

quic-garvgupt commented on clang/lib/Driver/ToolChains/BareMetal.cpp:
```diff
@@ -110,56 +144,99 @@ static std::string computeBaseSysRoot(const Driver &D, bool IncludeTriple) {
+  }
 
-  return Triple.getEnvironmentName() == "elf";
+  if (llvm::sys::fs::exists(SysRootDir))
```
Done

quic-garvgupt commented on clang/lib/Driver/ToolChains/BareMetal.cpp:
```diff
@@ -110,56 +144,99 @@ static std::string computeBaseSysRoot(const Driver &D, bool IncludeTriple) {
-  return Triple.isPPC() && Triple.getOS() == llvm::Triple::UnknownOS &&
-         Triple.getEnvironment() == llvm::Triple::EABI;
+// GCC mutltilibs will only work for those targets that have their multlib
+// structure encoded into GCCInstallation. Baremetal toolchain support ARM,
```
Done

quic-garvgupt commented:
> I've done another review pass. Mostly some small stylistic comments.
> 
> I'm finding it difficult to follow the code with the various different code-paths with similar names. I've made some suggestions on renaming. We may also be able to add some more comments to make it clearer.

I've made few more refactoring changes in addition to suggestion to improve code readability and facilitate the review process:
- Renamed the function from `hasGCCToolchainAlongSideClang` to `detectGCCToolchainAdjacent`, and added a comment explaining its purpose. (Pls let me know if there is need for a `fixme` or `todo` here to remove/fix this logic in future.)
- In the `computeSysRoot` function, renamed the local variable `SysRootDir` to `inferredSysRoot` for better clarity, and added additional comments to explain the logic.
- In the constructor, grouped all file path additions together and all program path additions together, with comments explaining the rationale behind each block.

quic-garvgupt commented on clang/lib/Driver/ToolChains/BareMetal.cpp:
```diff
@@ -110,56 +144,99 @@ static std::string computeBaseSysRoot(const Driver &D, bool IncludeTriple) {
+                     const ArgList &Args)
+    : Generic_ELF(D, Triple, Args) {
+  GCCInstallation.init(Triple, Args);
+  SysRoot = computeSysRoot();
```
> It is possible we may need to cache the SysRoot here. Assuming computeSysRoot is idempotent it can be done later.

Can you expand a bit on where do we need to cache the sysroot. I couldn't understood this comment clearly.

quic-garvgupt commented:
Hi @petrhosek, while petrsmith is reviewing this PR, if you don’t have any further suggestions, comments, or feedback, could you please provide an LGTM? That would help us proceed with merging the PR promptly once petr approves it. Thanks so much!

smithp35 reviewed (approved):
Thanks for the updates. I've marked it approved on my side. As mentioned before you'll want to get Petr Hosek or someone from his team to approve on the RISCV side as I can't comment on that part.

smithp35 reviewed (changes requested):
Apologies, after further testing I think I've found a problem that can break existing toolchains.

smithp35 commented on clang/lib/Driver/ToolChains/BareMetal.cpp:
```diff
@@ -110,56 +145,113 @@ static std::string computeBaseSysRoot(const Driver &D, bool IncludeTriple) {
+BareMetal::BareMetal(const Driver &D, const llvm::Triple &Triple,
+                     const ArgList &Args)
+    : Generic_ELF(D, Triple, Args) {
+  GCCInstallation.init(Triple, Args);
```
Thanks for the update. I'll rebuild the toolchain to check everything is still working.

smithp35 reviewed (approved):
I've rechecked the toolchain with the latest change and I can now build with `--target=arm-none-eabi -march=armv7-m`  with a GCC toolchain in `/usr/lib/arm-none-eabi` 

Putting the approve back on.

quic-garvgupt commented:
Rebased the stack on the tip of the branch. Will start merging the stack as soon as CI checks gets completed. Thanks again petrhosek and petersmith for reviewing the changes.

mysterymath commented:
This is consistently breaking the Fuchsia builders as seen in the above buildbot. Accordingly, I'm issuing a rollback.

lenary commented:
Do the fuchsia builders actually produce output on a failure? We seem to just get exit code 1, but no indication of what filecheck had problems with.

mysterymath commented:
> Do the fuchsia builders actually produce output on a failure? We seem to just get exit code 1, but no indication of what filecheck had problems with.

They do; here's a link! Let me know if you have an problems with the Web UI; I can paste things here if so.
https://luci-milo.appspot.com/ui/p/fuchsia/builders/toolchain.ci/clang-linux-x64/b8711783358287375697/overview

lenary commented:
Nice, I can navigate that UI.

Looks like a similar thing that I fixed-forward, but this time with a different unwind library, but clang isn't very happy with mixing unwind libraries anyway.

I think I'll leave this to @quic-garvgupt to fix, using the info from your link.

quic-garvgupt commented:
Fix tests failing on fuchsia clang x86_64 builders - https://github.com/llvm/llvm-project/pull/144655

quic-garvgupt commented:
> > Do the fuchsia builders actually produce output on a failure? We seem to just get exit code 1, but no indication of what filecheck had problems with.
> 
> They do; here's a link! Let me know if you have an problems with the Web UI; I can paste things here if so. https://luci-milo.appspot.com/ui/p/fuchsia/builders/toolchain.ci/clang-linux-x64/b8711783358287375697/overview

The issue doesn't just exist with expected string not found in standard input. Fuchsia sets CLANG_DEFAULT_UNWINDLIB to libunwind. As a result, when rtlib is set to libgcc and unwindlib is not explicitly specified, tests using Fuchsia as the default platform will error out with message `clang: error: --rtlib=libgcc requires --unwindlib=libgcc`.  This would either require tests to be xfailed or marked unsupported for fuchsia. I will include the fix for this in the reland patch itself.

petrhosek commented:
> > > Do the fuchsia builders actually produce output on a failure? We seem to just get exit code 1, but no indication of what filecheck had problems with.
> > 
> > 
> > They do; here's a link! Let me know if you have an problems with the Web UI; I can paste things here if so. https://luci-milo.appspot.com/ui/p/fuchsia/builders/toolchain.ci/clang-linux-x64/b8711783358287375697/overview
> 
> The issue doesn't just exist with expected string not found in standard input. Fuchsia sets CLANG_DEFAULT_UNWINDLIB to libunwind. As a result, when rtlib is set to libgcc and unwindlib is not explicitly specified, tests using Fuchsia as the default platform will error out with message `clang: error: --rtlib=libgcc requires --unwindlib=libgcc`. This would either require tests to be xfailed or marked unsupported for fuchsia. I will include the fix for this in the reland patch itself.

This issue is not specific to Fuchsia as a target, these builders build the Fuchsia Clang toolchain which is simply a distribution of Clang akin to other distributions such as Arm LLVM Toolchain. Using `XFAIL` isn't a valid solution since these tests aren't failing on Fuchsia. Rather, the tests needs to be written in a way that correctly handles different settings of `CLANG_DEFAULT_RTLIB` and `CLANG_DEFAULT_UNWINDLIB` which we've already done in different tests, typically by explicitly specifying `--rtlib=platform` and `--unwindlib=platform`.
